lib/Targets/AIETargetAirbin.cpp

```diff
@@ -0,0 +1,1746 @@
+//===- AIETargetAirbin.cpp --------------------------------------*- C++ -*-===//
+//
+// This file is licensed under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+// (c) Copyright 2021 Xilinx Inc.
+//
+//===----------------------------------------------------------------------===//
+
+#include "llvm/ADT/None.h"
+#include "llvm/Support/Format.h"
+#include "llvm/Support/raw_ostream.h"
+
+#include <algorithm> // find_if
+#include <array>
+#include <climits>
+#include <cstdint>
+#include <elf.h>
+#include <fcntl.h> // open
+#include <sstream>
+#include <type_traits> // enable_if_t
+#include <unistd.h>    // read
+#include <utility>     // pair
+#include <vector>
+
+#include "aie/AIENetlistAnalysis.h"
+#include "aie/Dialect/AIE/IR/AIEDialect.h"
+
+#include "AIETargets.h"
+
+// This target is a "flattening" of AIETargetXAIEV1 thru libXAIE.
+// All recorded writes are time/order invariant.
+// This allows sorting to compact the airbin,
+// but translating from XAIE is more difficult,
+// as some writes are handled by the runtime that loads our resulting airbin.
+
+// Current conventions for this file
+// - Some comments are code snippets from AIETargetXAIEV1.cpp.
+//   These serve as a guide for implementation
+
+// Marks a particular code path as unfinished.
+#define TODO assert(false)
+
+// Marks a particular code path as unused in normal execution.
+#define UNREACHABLE assert(false)
+
+namespace xilinx {
+namespace AIE {
+
+static constexpr auto disable = 0u;
+static constexpr auto enable = 1u;
+static constexpr auto MAX_CHANNEL_COUNT = 4u;
+
+static constexpr auto TILE_ADDR_OFF_WITDH = 18u;
+
+static constexpr auto TILE_ADDR_ROW_SHIFT = TILE_ADDR_OFF_WITDH;
+static constexpr auto TILE_ADDR_ROW_WIDTH = 5u;
+
+static constexpr auto TILE_ADDR_COL_SHIFT =
+    TILE_ADDR_ROW_SHIFT + TILE_ADDR_ROW_WIDTH;
+static constexpr auto TILE_ADDR_COL_WIDTH = 7u;
+
+static constexpr auto TILE_ADDR_ARR_SHIFT =
+    TILE_ADDR_COL_SHIFT + TILE_ADDR_COL_WIDTH;
+
```

Comment:
Perhaps this magic stuff should come out of libXAIE?

Reply:
Agreed, it would be best if the architecture specific values were factored out of all of the AIETargets